includes/embeds/class-amp-tumblr-embed-handler.php

```diff
@@ -6,56 +6,100 @@
+	 */
+	protected $base_embed_url = 'https://embed.tumblr.com/embed/post/';
+
 	/**
 	 * Register embed.
 	 */
 	public function register_embed() {
-		add_filter( 'embed_oembed_html', [ $this, 'filter_embed_oembed_html' ], 10, 2 );
+		// Not implemented.
 	}
 
 	/**
 	 * Unregister embed.
 	 */
 	public function unregister_embed() {
-		remove_filter( 'embed_oembed_html', [ $this, 'filter_embed_oembed_html' ], 10 );
+		// Not implemented.
 	}
 
 	/**
-	 * Filter oEmbed HTML for Tumblr to prepare it for AMP.
+	 * Sanitizes Tumblr raw embeds to make them AMP compatible.
 	 *
-	 * @param string $cache Cache for oEmbed.
-	 * @param string $url   Embed URL.
-	 * @return string Embed.
+	 * @param Document $dom DOM.
 	 */
-	public function filter_embed_oembed_html( $cache, $url ) {
-		$parsed_url = wp_parse_url( $url );
-		if ( false === strpos( $parsed_url['host'], 'tumblr.com' ) ) {
-			return $cache;
+	public function sanitize_raw_embeds( Document $dom ) {
+		$nodes = $dom->xpath->query( sprintf( '//div[ @class = "tumblr-post" and starts-with( @data-href, "%s" ) ]', $this->base_embed_url ) );
+
+		if ( 0 === $nodes->length ) {
+			return;
 		}
 
-		// @todo The iframe will not get sized properly.
-		if ( preg_match( '#data-href="(?P<href>https://embed.tumblr.com/embed/post/\w+/\w+)"#', $cache, $matches ) ) {
-			$cache = AMP_HTML_Utils::build_tag(
+		foreach ( $nodes as $node ) {
+			$iframe_src = $node->getAttribute( 'data-href' );
+
+			$attributes = [
+				'src'       => $iframe_src,
+				'layout'    => 'responsive',
+				'width'     => $this->args['width'],
+				'height'    => $this->args['height'],
+				'resizable' => '',
+				'sandbox'   => 'allow-scripts allow-popups allow-same-origin',
+			];
+
+			$amp_node = AMP_DOM_Utils::create_node(
+				$dom,
 				'amp-iframe',
+				$attributes
+			);
+
+			// Add an overflow node to allow the amp-iframe to resize.
+			$overflow_node              = AMP_DOM_Utils::create_node(
+				$dom,
+				'div',
 				[
-					'width'   => $this->args['width'],
-					'height'  => $this->args['height'],
-					'layout'  => 'responsive',
-					'sandbox' => 'allow-scripts allow-popups', // The allow-scripts is needed to allow the iframe to render; allow-popups needed to allow clicking.
-					'src'     => $matches['href'],
-				],
-				sprintf( '<a placeholder href="%s">Tumblr</a>', $url )
+					'overflow'   => '',
+					'tabindex'   => 0,
+					'role'       => 'button',
+					'aria-label' => __( 'See more', 'amp' ),
+				]
 			);
-		}
+			$overflow_node->textContent = __( 'See more', 'amp' );
+			$amp_node->appendChild( $overflow_node );
+
+			// Append the original link as a placeholder node.
+			if ( $node->firstChild instanceof DOMElement && 'a' === $node->firstChild->nodeName ) {
+				$placeholder_node = $node->firstChild;
+				$placeholder_node->setAttribute( 'placeholder', '' );
+				$amp_node->appendChild( $placeholder_node );
+			}
+
+			$this->maybe_remove_script_sibling( $node, 'assets.tumblr.com/post.js' );
```

Comment:
Per my comment above, this could then be changed to be:

```suggestion
			$this->maybe_remove_script_sibling(
				$node,
				function ( DOMElement $script ) {
					if ( ! $script->hasAttribute( 'src' ) ) {
						return false;
					}
					$parsed_url = wp_parse_url( $script->getAttribute( 'src' ) );
					return (
						( isset( $parsed_url['host'], $parsed_url['path'] ) )
						&&
						'assets.tumblr.com' === $parsed_url['host']
						&&
						'/post.js' === $parsed_url['path']
					);
				}
			);
```